src/Backups/BackupsWorker.cpp

```diff
@@ -744,32 +733,108 @@ std::vector<BackupsWorker::Info> BackupsWorker::getAllActiveRestoreInfos() const
     return res_infos;
 }
 
-bool BackupsWorker::hasConcurrentBackups(const BackupSettings & backup_settings) const
+bool BackupsWorker::hasConcurrentBackups(const OperationID & backup_id, const ContextPtr & context, bool on_cluster) const
 {
-    /// Check if there are no concurrent backups
-    if (num_active_backups)
+    if (on_cluster)
     {
-        /// If its an internal backup and we currently have 1 active backup, it could be the original query, validate using backup_uuid
-        if (!(num_active_backups == 1 && backup_settings.internal && getAllActiveBackupInfos().at(0).id == toString(*backup_settings.backup_uuid)))
+        String common_backup_path = context->getConfigRef().getString("backups.zookeeper_path", "/clickhouse/backups") ;
+        auto zookeeper = context->getGlobalContext()->getZooKeeper();
+        std::string backup_stage_path = common_backup_path + "/backup-" + toString(backup_id) +"/stage";
+
+        if (!zookeeper->exists(common_backup_path))
+            zookeeper->createAncestors(common_backup_path);
+
+        for (size_t attempt = 0; attempt < MAX_ZOOKEEPER_ATTEMPTS; ++attempt)
         {
-            return true;
+            Coordination::Stat stat;
+            zookeeper->get(common_backup_path, &stat);
+            Strings existing_backup_paths = zookeeper->getChildren(common_backup_path);
+
+            for (const auto & existing_backup_path : existing_backup_paths)
+            {
+                if (startsWith(existing_backup_path, "restore-"))
+                    continue;
```

Comment:
I'd prefer not to overload BackupsWorker with logic like this. Let's move it to `BackupCoordinationRemove::hasConcurrentBackups()` ? We can also add `BackupCoordinationLocal::hasConcurrentBackups()` which will just check whether only one instance of `BackupCoordinationLocal` exists.

Reply:
addressed it.